pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -380,25 +392,47 @@ func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
 // cm.UpdatePluginResource() run during predicate Admit guarantees we adjust nodeinfo
 // capacity for already allocated pods so that they can continue to run. However, new pods
 // requiring device plugin resources will not be scheduled till device plugin re-registers.
-func (m *ManagerImpl) GetCapacity() (v1.ResourceList, []string) {
+func (m *ManagerImpl) GetCapacity() (v1.ResourceList, v1.ResourceList, []string) {
 	needsUpdateCheckpoint := false
 	var capacity = v1.ResourceList{}
+	var allocatable = v1.ResourceList{}
 	var deletedResources []string
 	m.mutex.Lock()
-	for resourceName, devices := range m.allDevices {
+	for resourceName, devices := range m.healthyDevices {
 		if _, ok := m.endpoints[resourceName]; !ok {
-			delete(m.allDevices, resourceName)
+			delete(m.healthyDevices, resourceName)
 			deletedResources = append(deletedResources, resourceName)
 			needsUpdateCheckpoint = true
 		} else {
 			capacity[v1.ResourceName(resourceName)] = *resource.NewQuantity(int64(devices.Len()), resource.DecimalSI)
+			allocatable[v1.ResourceName(resourceName)] = *resource.NewQuantity(int64(devices.Len()), resource.DecimalSI)
+		}
+	}
+	for resourceName, devices := range m.unhealthyDevices {
+		if _, ok := m.endpoints[resourceName]; !ok {
+			delete(m.unhealthyDevices, resourceName)
+			alreadyDeleted := false
+			for _, name := range deletedResources {
+				if name == resourceName {
+					alreadyDeleted = true
+				}
+			}
+			if !alreadyDeleted {
+				deletedResources = append(deletedResources, resourceName)
+			}
+			needsUpdateCheckpoint = true
+		} else {
+			capacityCount := capacity[v1.ResourceName(resourceName)]
+			unhealthyCount := *resource.NewQuantity(int64(devices.Len()), resource.DecimalSI)
+			capacityCount.Add(unhealthyCount)
```

Comment:
What if resource doesn't exist in capacity yet? Would we get a segfault here? Can we add some unit test for this?

Reply:
Test is already there: https://github.com/vikaschoudhary16/kubernetes/blob/dc541fa0365de94d1c201baa74bce84d20bd9553/pkg/kubelet/cm/deviceplugin/manager_test.go#L203-L211

Looks like it wont crash. Just take a look here:
https://play.golang.org/p/pi1bI3caJRL

Reply:
Interesting to know. I guess the map lookup just returns struct with zero values when the key doesn't exist.